File: algorithms/fibonacci/julia/fibonacci.jl
function fibArr(n, s=Array{BigInt}([0,1]))
    #=
    (int, arr) -> arr
    =#
    # base case: return the array with n elements
    if length(s) > n
        return s[1:n+1]
    end
    # rcursive case: Add one more element
    return fibArr(n, append!(s, [s[end] + s[end-1]]))
end

memo = Dict{Int64, BigInt}(0=>0, 1=>1)
function fib(n)
    # remember every number that is different
    if !(haskey(memo, n))
        memo[n] = fib(n-1) + fib(n-2)
    end
    # contained element
    return memo[n]
end

# println(fibArr(0))
# println(fibArr(1))
# println(fibArr(2))
# println(fibArr(5))
# println(fibArr(10))
# println(fibArr(20))

# println(fib(0))
# println(fib(1))
# println(fib(2))
# println(fib(5))
# println(fib(10))
# println(fib(20))

# Dealing with big numbers
# println(fibArr(BigInt(100)))
# println(fib(BigInt(100)))

File: algorithms/fibonacci/python/fibonacci.py
def fibArr(n, s=[0,1]):
    """(int, arr) -> arr
    Returns an array containing the first n+1 fibonacci numbers.
    This function is computed recursively.
    """
    # return the n elements in an array
    if (len(s) > n):
        return s[0:n+1]
    # add the next number to the sequence
    return fibArr(n, s+[s[-2] + s[-1]])


from functools import lru_cache
@lru_cache(maxsize=None)
def fib(n):
    """(int) -> int
    Returns an integer, being the nth fibonacci number.
    This function is computed recursively and make use of
    the decorator @functools.lru_cache().
    """
    # base case
    if n < 2:
        return n
    # recursive case
    return fib(n-1) + fib(n-2)



# print(fibArr(0))
# print(fibArr(1))
# print(fibArr(2))
# print(fibArr(5))
# print(fibArr(10))
# print(fibArr(20))

# print(fib(0))
# print(fib(1))
# print(fib(2))
# print(fib(5))
# print(fib(10))
# print(fib(20))
